File: security-platform/agents/dotnet/SecurityPlatform.Agent/Redactor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace SecurityPlatform.Agent
{
    /// <summary>
    /// Redacts PII from data structures.
    /// </summary>
    public class Redactor
    {
        private const string DefaultReplacement = "***REDACTED***";
        private readonly List<RedactionRule> _rules;

        public Redactor(List<RedactionRule> rules)
        {
            _rules = rules ?? new List<RedactionRule>();
        }

        public object? Redact(object? data)
        {
            if (data == null)
                return null;

            return data switch
            {
                Dictionary<string, object> dict => RedactDictionary(dict),
                IDictionary<string, object> dict => RedactDictionary(dict.ToDictionary(kvp => kvp.Key, kvp => kvp.Value)),
                System.Collections.IList list => RedactList(list),
                string str => RedactString(str),
                _ => data
            };
        }

        private Dictionary<string, object> RedactDictionary(Dictionary<string, object> dict)
        {
            var result = new Dictionary<string, object>();
            foreach (var kvp in dict)
            {
                var key = kvp.Key;
                var value = kvp.Value;

                bool shouldRedact = _rules.Any(rule =>
                    (!string.IsNullOrEmpty(rule.Field) && rule.Field.Equals(key, StringComparison.OrdinalIgnoreCase)) ||
                    (!string.IsNullOrEmpty(rule.Pattern) && Regex.IsMatch(key, rule.Pattern, RegexOptions.IgnoreCase)));

                result[key] = shouldRedact ? DefaultReplacement : Redact(value);
            }
            return result;
        }

        private Dictionary<string, object> RedactDictionary(IDictionary<string, object> dict)
        {
            return RedactDictionary(dict.ToDictionary(kvp => kvp.Key, kvp => kvp.Value));
        }

        private System.Collections.IList RedactList(System.Collections.IList list)
        {
            var result = new List<object?>();
            foreach (var item in list)
            {
                result.Add(Redact(item));
            }
            return result;
        }

        private string RedactString(string str)
        {
            foreach (var rule in _rules)
            {
                if (!string.IsNullOrEmpty(rule.Pattern))
                {
                    var replacement = rule.Replacement ?? DefaultReplacement;
                    str = Regex.Replace(str, rule.Pattern, replacement, RegexOptions.IgnoreCase);
                }
            }
            return str;
        }
    }
}

File: security-platform/agents/ios/SecurityPlatform/Redactor.swift
import Foundation

class Redactor {
    private let rules: [RedactionRule]
    private static let defaultReplacement = "***REDACTED***"

    init(rules: [RedactionRule]) {
        self.rules = rules
    }

    func redact(_ data: Any) -> Any {
        if let dict = data as? [String: Any] {
            return redactDictionary(dict)
        } else if let array = data as? [Any] {
            return array.map { redact($0) }
        } else if let str = data as? String {
            return redactString(str)
        }
        return data
    }

    private func redactDictionary(_ dict: [String: Any]) -> [String: Any] {
        var result: [String: Any] = [:]

        for (key, value) in dict {
            var shouldRedact = false

            for rule in rules {
                if let field = rule.field, field.lowercased() == key.lowercased() {
                    shouldRedact = true
                    break
                }
                if let pattern = rule.pattern {
                    let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive)
                    if let range = NSRange(key, in: key) as NSRange?,
                       regex?.firstMatch(in: key, range: range) != nil {
                        shouldRedact = true
                        break
                    }
                }
            }

            if shouldRedact {
                let replacement = rules.first(where: { $0.field == key || $0.pattern != nil })?.replacement ?? Self.defaultReplacement
                result[key] = replacement
            } else {
                result[key] = redact(value)
            }
        }

        return result
    }

    private func redactString(_ str: String) -> String {
        var result = str
        for rule in rules {
            if let pattern = rule.pattern {
                let replacement = rule.replacement ?? Self.defaultReplacement
                if let regex = try? NSRegularExpression(pattern: pattern, options: .caseInsensitive) {
                    result = regex.stringByReplacingMatches(in: result,
                                                             options: [],
                                                             range: NSRange(location: 0, length: result.utf16.count),
                                                             withTemplate: replacement)
                }
            }
        }
        return result
    }
}
